### PR TITLE
Al/scale

### PR DESCRIPTION
**Pull request by alundgard**
Refactor Scale.ts, scaleActions.ts, scalesReducer.ts

### CODE REVIEW

File: src/js/reducers/scalesReducer.ts [hunk 1]
```diff
@@ -1,30 +1,34 @@
...
+
+function scalesReducer(state: ScaleState, action) {
   if (typeof state === 'undefined') {
-    return Immutable.Map();
+    return Map();
   }
 
   if (action.type === ACTIONS.ADD_SCALE) {
-    return set(state, action.id, Immutable.fromJS(action.props));
+    return set(state, action.id, fromJS(action.props));
```
Comment (alundgard) [inline src/js/reducers/scalesReducer.ts]: state.set

File: src/js/actions/scaleActions.ts [hunk 2]
```diff
@@ -1,70 +1,23 @@
...
-    property: property,
-    ref: ref
-  };
-}
+// /**
+//  * Action creator to create a new scale and add it to the store.
+//  *
+//  */
+export const addScale = createStandardAction('ADD_SCALE').map((payload: ScaleRecord) => {
+  const id = dl.extend({
```
Comment (alundgard) [inline src/js/actions/scaleActions.ts]: remove dl.extend

File: src/js/reducers/scalesReducer.ts [hunk 3]
```diff
@@ -9,34 +9,32 @@ const setIn = immutableUtils.setIn;
...
-  if (action.type === ACTIONS.UPDATE_SCALE_PROPERTY) {
+  if (action.type === getType(scaleActions.updateScaleProperty)) {
     return setIn(state, action.id + '.' + action.property,
       fromJS(action.value));
   }
 
-  if (action.type === ACTIONS.AMEND_DATA_REF) {
+  if (action.type === getType(scaleActions.amendDataRef)) {
     const refs = getIn(state, action.id + '.' + action.property);
     return setIn(state, action.id + '.' + action.property, refs.push(action.ref));
```
Comment (alundgard) [inline src/js/reducers/scalesReducer.ts]: state.setIn

File: src/js/store/factory/Scale.ts [hunk 4]
```diff
@@ -1,48 +1,22 @@
 import {Map, Record, RecordOf} from 'immutable';
+import {Scale as ScaleType} from 'vega-typings/types';
 
 /**
  * Scales are functions that transform a domain of data values (numbers, dates, strings, etc.) to a range of visual values (pixels, colors, sizes).
  */
-export interface LyraScale {
+export interface _LyraScale {
```
Comment (alundgard) [inline src/js/store/factory/Scale.ts]: export type
LyraScale

File: src/js/actions/scaleActions.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+// /**
+//  * Action creator to create a new scale and add it to the store.
+//  */
+export const addScale = createStandardAction('ADD_SCALE').map((payload: ScaleRecord) => {
+  const id = payload._id || counter.global();
+  return {payload: payload.merge({_id: id}), meta: id}
+});
+
+export const updateScaleProperty = createStandardAction('UPDATE_SCALE_PROPERTY')<{scaleId: number, property: any, value: any}, number>();
```
Comment (jonathanzong) [inline src/js/actions/scaleActions.ts]: `scaleId` is `meta`, so it doesn't need to be part of the `payload` too

```
export const updateScaleProperty = createStandardAction('UPDATE_SCALE_PROPERTY')<{property: any, value: any}, number>();
```

File: src/js/actions/scaleActions.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+export const amendDataRef = createStandardAction('AMEND_DATA_REF')<{scaleId: number, property: any, ref: any}, number>();
```
Comment (jonathanzong) [inline src/js/actions/scaleActions.ts]: see above

File: src/js/actions/scaleActions.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+export const deleteScale = createStandardAction('DELETE_SCALE')<{id: number}, number>();
```
Comment (jonathanzong) [inline src/js/actions/scaleActions.ts]: `<null, number>`

File: src/js/reducers/scalesReducer.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  if (action.type === getType(scaleActions.amendDataRef)) {
+    const p = action.payload;
+    const refs = state.getIn([str(id), p.property]);
+    return state.setIn([str(id), p.property], refs.push(p.ref));
+  }
+
+  if (action.type === getType(scaleActions.deleteScale)) {
+    return state.delete(str(id));
```
Comment (jonathanzong) [inline src/js/reducers/scalesReducer.ts]: `delete` isn't a function on records, you want `remove`

File: src/js/actions/scaleActions.ts (hunk 5 above)
Comment (jonathanzong) [reply src/js/actions/scaleActions.ts]: i wanna say `property` is a `string`

File: src/js/store/factory/Scale.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  _domain: [],
+  _range: [],
+
+} & ScaleType;
+
+const names = {};
+
+// Scales churn (unused scales are deleted) and thus we want to reuse names
+// as much as possible.
+function rename(name) {
```
Comment (jonathanzong) [inline src/js/store/factory/Scale.ts]: `function rename(name: string): string {`

just to be safe

**Comment by jonathanzong** [comment]
this looks good to go after comments, you'll also need to update references to `Scale` and especially `scaleActions` because the API for the action creators have changed, but can do that in a separate PR

**Pull request merged**